File: mergeSort.cpp
#include <iostream>
#include <vector>

void merge(std::vector<int>& arr, int l, int mid, int r) {
    int an = mid - l + 1;
    int bn = r - (mid + 1) + 1;
    std::vector<int> a(an); 
    std::vector<int> b(bn); 

    for (int i = 0; i < an; i++) {
        a[i] = arr[i + l]; 
    }
    for (int j = 0; j < bn; j++) {
        b[j] = arr[mid + 1 + j]; 
    }
    int i = 0;
    int j = 0;
    int k = l;
    while (i < an && j < bn) {
        if (a[i] < b[j]) {
            arr[k++] = a[i++];
        } else
            arr[k++] = b[j++];
    }
    while (i < an) {
        arr[k++] = a[i++];
    }
    while (j < bn) {

        arr[k++] = b[j++];
    }
    return;
}

void mergesort(std::vector<int>& arr, int l, int r) {
    if (l >= r)
        return;
    {
        int mid = (l + r) / 2;
        mergesort(arr, l, mid);
        mergesort(arr, mid + 1, r);
        merge(arr, l, mid, r);

        return;
    }
}

int main() {
    std::vector<int> arr = {8, 5, 4, 7, 6};
    int size = arr.size();
    mergesort(arr, 0, size - 1);
    std::cout << "The sorted array: ";
    for (int i = 0; i < size; i++) {
        std::cout << arr[i] << " ";
    }

    return 0;
}

File: MergeSort.js
function merge(arr, l, mid, r) {
    let an = mid - l + 1;
    let bn = r - (mid + 1) + 1;
    let a = arr.slice(l, mid + 1); // create one array
    let b = arr.slice(mid + 1, r + 1); // create another array

    let i = 0;
    let j = 0;
    let k = l;
    while (i < an && j < bn) {
        if (a[i] < b[j]) {
            arr[k++] = a[i++];
        } else {
            arr[k++] = b[j++];
        }
    }
    while (i < an) {
        arr[k++] = a[i++];
    }
    while (j < bn) {
        arr[k++] = b[j++];
    }
}

function mergesort(arr, l, r) {
    if (l >= r) return;

    let mid = Math.floor((l + r) / 2);
    mergesort(arr, l, mid);
    mergesort(arr, mid + 1, r);
    merge(arr, l, mid, r);
}

let arr = [8, 5, 4, 7, 6];
let size = arr.length;
mergesort(arr, 0, size - 1);
console.log("The sorted array:", arr.join(" "));
